src/python/pants/backend/project_info/depgraph_test.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from textwrap import dedent
+
+import pytest
+
+from pants.backend.project_info.depgraph import (
+    DependencyGraph,
+    DependencyGraphRequest,
+    Edge,
+    Granularity,
+    Vertex,
+    rules,
+)
+from pants.backend.python.target_types import PythonLibrary, PythonRequirementLibrary
+from pants.engine.rules import QueryRule
+from pants.engine.target import SpecialCasedDependencies, Target
+from pants.testutil.rule_runner import RuleRunner
+
+
+# We verify that any subclasses of `SpecialCasedDependencies` will show up with the `dependencies`
+# goal by creating a mock target.
+class SpecialDepsField(SpecialCasedDependencies):
+    alias = "special_deps"
+
+
+class SpecialDepsTarget(Target):
+    alias = "special_deps_tgt"
+    core_fields = (SpecialDepsField,)
+
+
+@pytest.fixture
+def rule_runner() -> RuleRunner:
+    return RuleRunner(
+        rules=[*rules(), QueryRule(DependencyGraph, [DependencyGraphRequest])],
+        target_types=[PythonLibrary, PythonRequirementLibrary, SpecialDepsTarget],
+    )
+
+
+def test_depgraph_construction(rule_runner: RuleRunner) -> None:
+    rule_runner.add_to_build_file(
+        "3rdparty/python",
+        dedent(
+            """\
+            python_requirement_library(
+                name='extlib',
+                requirements=['extlib==1.2.3'],
+            )
+            """
+        ),
+    )
+    rule_runner.add_to_build_file(
+        "src/python/foo/",
+        dedent(
+            """\
+            python_library(dependencies=['3rdparty/python:extlib'])
+            """
+        ),
+    )
+    rule_runner.create_file("src/python/foo/bar.py", "")
+    rule_runner.add_to_build_file(
+        "src/python/baz/",
+        dedent(
+            """\
+            python_library(name="lib", dependencies=['src/python/foo'])
+            """
+        ),
+    )
+    rule_runner.create_file("src/python/baz/qux.py", "")
```

Comment:
I've been trying very actively to drop `add_to_build_file` and `create_file` in favor of the batched `write_files()`. Not blocking, but you save me future work by using`.write_files()` when adding new code.

Reply:
Why though?

Reply:
Because it batches all setup of files, which makes it easier to see what's happening. And 1 consistent, simple API > 2 slightly different APIs (one appends, other creates; one makes you use the whole file name, other has you leave off BUILD)